components/org.wso2.carbon.identity.application.authentication.handler.identifier/src/main/java/org/wso2/carbon/identity/application/authentication/handler/identifier/UserResolveExecutor.java

```diff
@@ -0,0 +1,201 @@
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.wso2.carbon.identity.flow.execution.engine.Constants.ExecutorStatus.STATUS_ERROR;
+import static org.wso2.carbon.identity.flow.execution.engine.Constants.ExecutorStatus.STATUS_USER_INPUT_REQUIRED;
+import static org.wso2.carbon.identity.flow.execution.engine.Constants.ExecutorStatus.STATUS_COMPLETE;
+import static org.wso2.carbon.identity.flow.execution.engine.Constants.USERNAME_CLAIM_URI;
+
+/**
+ * This class is responsible for resolving user.
+ */
+public class UserResolveExecutor implements Executor {
+
+    public static final String USER_RESOLVE_EXECUTOR = "UserResolveExecutor";
+    private static final Log log = LogFactory.getLog(UserResolveExecutor.class);
+
+    /**
+     * Returns the name of the executor.
+     *
+     * @return Name of the executor.
+     */
+    public String getName() {
+
+        return USER_RESOLVE_EXECUTOR;
+    }
+
+    /**
+     * Executes the user resolution logic. If the username claim is not present, signals that user input is required.
+     * Otherwise, resolves user attributes and adds them to the flow context.
+     *
+     * @param context Flow execution context.
+     * @return ExecutorResponse indicating the status of execution.
+     * @throws FlowEngineException If an error occurs during execution.
+     */
+    @Override
+    public ExecutorResponse execute(FlowExecutionContext context) throws FlowEngineException {
+
+        ExecutorResponse executorResponse;
+        String usernameClaim = (String) context.getFlowUser().getClaim(FrameworkConstants.USERNAME_CLAIM);
+        if (usernameClaim == null) {
+            executorResponse = new ExecutorResponse(STATUS_USER_INPUT_REQUIRED);
+        } else {
+            String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(usernameClaim);
+            executorResponse = resolveUser(tenantAwareUsername, context.getTenantDomain(), context);
+        }
+        return executorResponse;
+    }
+
+    /**
+     * Resolves user attributes and adds them to the flow context.
+     *
+     * @param username     Username to resolve.
+     * @param tenantDomain Tenant domain.
+     * @param context      Flow execution context.
+     */
+    private ExecutorResponse resolveUser(String username, String tenantDomain, FlowExecutionContext context) {
+
+        ExecutorResponse executorResponse;
+        try {
+            // Obtain the realm service and user realm for the tenant.
+            RealmService realmService = IdentifierAuthenticatorServiceComponent.getRealmService();
+            int tenantId = realmService.getTenantManager().getTenantId(tenantDomain);
+            UserRealm userRealm = (UserRealm) realmService.getTenantUserRealm(tenantId);
+
+            if (userRealm == null) {
+                throw new AuthenticationFailedException(
```

Comment:
let's rewrite this logic to handle the excpetions properly. IMO throwing such error doesn't make sense in this scenario.